In [82]:
import pandas as pd

In [83]:
df = pd.read_csv(r"C:\Users\HP\OneDrive\Documents\mutual-fund-analytics\data\raw\02_nav_history - Copy.csv")
df.head(5)

,amfi_code,date,nav
0,119551,2022-01-03,54.3856
1,119551,2022-01-04,54.3474
2,119551,2022-01-05,54.6869
3,119551,2022-01-06,55.4550
4,119551,2022-01-07,55.3692


In [84]:
df.shape

(46000, 3)

In [85]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 46000 entries, 0 to 45999
Data columns (total 3 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   amfi_code  46000 non-null  int64  
 1   date       46000 non-null  object 
 2   nav        46000 non-null  float64
dtypes: float64(1), int64(1), object(1)
memory usage: 1.1+ MB


In [86]:
df.isnull().sum()

amfi_code    0
date         0
nav          0
dtype: int64

In [87]:
df.duplicated().sum()

np.int64(0)

In [88]:
df.describe(include='all')

,amfi_code,date,nav
count,46000.000000,46000,46000.000000
unique,NaN,1150,NaN
top,NaN,2022-01-03,NaN
freq,NaN,40,NaN
mean,120247.000000,NaN,269.570265
std,14352.317221,NaN,577.187060
min,100016.000000,NaN,26.136600
25%,118632.750000,NaN,69.170425
50%,119551.500000,NaN,122.732150
75%,120842.250000,NaN,260.338675


In [89]:
df['date'] = pd.to_datetime(df['date'])

In [90]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 46000 entries, 0 to 45999
Data columns (total 3 columns):
 #   Column     Non-Null Count  Dtype         
---  ------     --------------  -----         
 0   amfi_code  46000 non-null  int64         
 1   date       46000 non-null  datetime64[ns]
 2   nav        46000 non-null  float64       
dtypes: datetime64[ns](1), float64(1), int64(1)
memory usage: 1.1 MB


In [91]:
df = df.sort_values(
    ['amfi_code', 'date']
).reset_index(drop=True)

In [92]:
df.head(5)

,amfi_code,date,nav
0,100016,2022-01-03,520.4608
1,100016,2022-01-04,515.0971
2,100016,2022-01-05,521.7239
3,100016,2022-01-06,515.7880
4,100016,2022-01-07,515.1639


In [93]:
df = df.drop_duplicates()

In [94]:
#No missing NAV values found.
#Forward-fill step not required.

In [95]:
df = df[df['nav'] > 0]

In [96]:
print("Minimum NAV:", df['nav'].min())

Minimum NAV: 26.1366


In [97]:
df.to_csv("../data/processed/nav_history_clean.csv", index=False)


# NAV History Data Cleaning

### Objective
Clean and validate historical NAV data for mutual fund schemes.

### Cleaning Steps Performed

- Converted `date` column to datetime format.
- Sorted records by `amfi_code` and `date`.
- Checked for missing values.
- Checked for duplicate records .
- Validated that all NAV values are greater than zero.
- Reviewed dataset quality for further analysis and database loading.

### Findings

- No missing values found in `amfi_code`, `date`, or `nav`.
- No invalid NAV values detected.
- Forward-fill step not required.
- Dataset successfully cleaned and prepared for downstream analysis.

### Output

Cleaned dataset saved as:

`data/processed/nav_history_clean.csv`

Scheme Performance

In [98]:
df2=pd.read_csv(r"C:\Users\HP\OneDrive\Documents\mutual-fund-analytics\data\raw\07_scheme_performance - Copy.csv")
df2.head(5)

,amfi_code,scheme_name,fund_house,category,plan,return_1yr_pct,return_3yr_pct,return_5yr_pct,benchmark_3yr_pct,alpha,beta,sharpe_ratio,sortino_ratio,std_dev_ann_pct,max_drawdown_pct,aum_crore,expense_ratio_pct,morningstar_rating,risk_grade
0,119551,SBI Bluechip Fund - Regular Plan - Growth,SBI Mutual Fund,Large Cap,Regular,12.42,12.36,14.45,11.49,0.87,0.89,0.88,1.29,14.0,-21.70,14288,1.54,4,Moderate
1,119552,SBI Bluechip Fund - Direct Plan - Growth,SBI Mutual Fund,Large Cap,Direct,15.25,11.30,14.23,9.52,1.78,0.87,0.81,1.29,14.0,-24.43,1231,0.66,3,Moderate
2,119598,SBI Small Cap Fund - Regular Plan - Growth,SBI Mutual Fund,Small Cap,Regular,24.56,23.39,20.67,22.16,1.23,0.89,0.94,1.35,25.0,-13.35,19259,1.43,5,Very High
3,119599,SBI Small Cap Fund - Direct Plan - Growth,SBI Mutual Fund,Small Cap,Direct,20.59,23.14,21.82,22.01,1.13,1.04,0.93,1.67,25.0,-24.78,36061,0.72,4,Very High
4,119120,SBI Magnum Gilt Fund - Regular Plan - Growth,SBI Mutual Fund,Gilt,Regular,5.34,6.07,5.43,4.47,1.60,0.22,1.52,2.11,4.0,-2.30,24101,0.77,5,Low


In [99]:
df2.shape

(40, 19)

In [100]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 40 entries, 0 to 39
Data columns (total 19 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   amfi_code           40 non-null     int64  
 1   scheme_name         40 non-null     object 
 2   fund_house          40 non-null     object 
 3   category            40 non-null     object 
 4   plan                40 non-null     object 
 5   return_1yr_pct      40 non-null     float64
 6   return_3yr_pct      40 non-null     float64
 7   return_5yr_pct      40 non-null     float64
 8   benchmark_3yr_pct   40 non-null     float64
 9   alpha               40 non-null     float64
 10  beta                40 non-null     float64
 11  sharpe_ratio        40 non-null     float64
 12  sortino_ratio       40 non-null     float64
 13  std_dev_ann_pct     40 non-null     float64
 14  max_drawdown_pct    40 non-null     float64
 15  aum_crore           40 non-null     int64  
 16  expense_ra

In [101]:
df2.isnull().sum()

amfi_code             0
scheme_name           0
fund_house            0
category              0
plan                  0
return_1yr_pct        0
return_3yr_pct        0
return_5yr_pct        0
benchmark_3yr_pct     0
alpha                 0
beta                  0
sharpe_ratio          0
sortino_ratio         0
std_dev_ann_pct       0
max_drawdown_pct      0
aum_crore             0
expense_ratio_pct     0
morningstar_rating    0
risk_grade            0
dtype: int64

In [102]:
df2.duplicated().sum()

np.int64(0)

In [103]:
df2.describe(include='all')

,amfi_code,scheme_name,fund_house,category,plan,return_1yr_pct,return_3yr_pct,return_5yr_pct,benchmark_3yr_pct,alpha,beta,sharpe_ratio,sortino_ratio,std_dev_ann_pct,max_drawdown_pct,aum_crore,expense_ratio_pct,morningstar_rating,risk_grade
count,40.000000,40,40,40,40,40.000000,40.000000,40.000000,40.000000,40.000000,40.000000,40.000000,40.000000,40.000000,40.000000,40.00000,40.000000,40.000000,40
unique,NaN,40,10,12,2,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,5
top,NaN,SBI Bluechip Fund - Regular Plan - Growth,SBI Mutual Fund,Large Cap,Regular,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,Moderate
freq,NaN,1,5,14,32,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,16
mean,120247.000000,NaN,NaN,NaN,NaN,14.376000,14.089000,14.516750,12.835500,1.253500,0.873250,1.361750,2.082500,14.962500,-19.200250,26091.60000,1.237000,4.250000,NaN
std,14534.998667,NaN,NaN,NaN,NaN,4.883023,4.617253,4.454021,4.740972,0.447412,0.224846,1.475805,2.203144,6.669282,8.819164,13809.11134,0.386584,0.742484,NaN
min,100016.000000,NaN,NaN,NaN,NaN,4.260000,5.140000,5.430000,3.960000,0.510000,0.220000,0.800000,1.030000,0.500000,-33.500000,979.00000,0.550000,3.000000,NaN
25%,118632.750000,NaN,NaN,NaN,NaN,11.735000,12.035000,12.340000,10.690000,0.887500,0.890000,0.865000,1.270000,14.000000,-25.062500,17400.50000,0.787500,4.000000,NaN
50%,119551.500000,NaN,NaN,NaN,NaN,14.620000,14.205000,14.185000,13.090000,1.205000,0.960000,0.925000,1.445000,14.000000,-20.600000,26713.00000,1.425000,4.000000,NaN
75%,120842.250000,NaN,NaN,NaN,NaN,16.392500,15.882500,17.585000,14.775000,1.700000,1.000000,0.985000,1.637500,19.000000,-14.255000,38125.00000,1.540000,5.000000,NaN


In [104]:
#Return columns were inspected and confirmed to be numeric (float64) with no missing values. No conversion was required.
#return_cols = [
#    'return_1yr_pct',
 #   'return_3yr_pct',
  #  'return_5yr_pct']


#for col in return_cols:
 #   df2[col] = pd.to_numeric(
  #      df2[col],
   #     errors='coerce')
    

In [105]:
df2[['return_1yr_pct',
     'return_3yr_pct',
     'return_5yr_pct']].describe()

,return_1yr_pct,return_3yr_pct,return_5yr_pct
count,40.000000,40.000000,40.000000
mean,14.376000,14.089000,14.516750
std,4.883023,4.617253,4.454021
min,4.260000,5.140000,5.430000
25%,11.735000,12.035000,12.340000
50%,14.620000,14.205000,14.185000
75%,16.392500,15.882500,17.585000
max,24.930000,23.390000,23.800000


In [106]:
df2['anomaly_flag'] = (
    (df2['return_1yr_pct'] > 50) |
    (df2['return_3yr_pct'] > 50) |
    (df2['return_5yr_pct'] > 50)
)

In [108]:
df2.head()

,amfi_code,scheme_name,fund_house,category,plan,return_1yr_pct,return_3yr_pct,return_5yr_pct,benchmark_3yr_pct,alpha,beta,sharpe_ratio,sortino_ratio,std_dev_ann_pct,max_drawdown_pct,aum_crore,expense_ratio_pct,morningstar_rating,risk_grade,anomaly_flag
0,119551,SBI Bluechip Fund - Regular Plan - Growth,SBI Mutual Fund,Large Cap,Regular,12.42,12.36,14.45,11.49,0.87,0.89,0.88,1.29,14.0,-21.70,14288,1.54,4,Moderate,False
1,119552,SBI Bluechip Fund - Direct Plan - Growth,SBI Mutual Fund,Large Cap,Direct,15.25,11.30,14.23,9.52,1.78,0.87,0.81,1.29,14.0,-24.43,1231,0.66,3,Moderate,False
2,119598,SBI Small Cap Fund - Regular Plan - Growth,SBI Mutual Fund,Small Cap,Regular,24.56,23.39,20.67,22.16,1.23,0.89,0.94,1.35,25.0,-13.35,19259,1.43,5,Very High,False
3,119599,SBI Small Cap Fund - Direct Plan - Growth,SBI Mutual Fund,Small Cap,Direct,20.59,23.14,21.82,22.01,1.13,1.04,0.93,1.67,25.0,-24.78,36061,0.72,4,Very High,False
4,119120,SBI Magnum Gilt Fund - Regular Plan - Growth,SBI Mutual Fund,Gilt,Regular,5.34,6.07,5.43,4.47,1.60,0.22,1.52,2.11,4.0,-2.30,24101,0.77,5,Low,False


In [109]:
df2['anomaly_flag'].value_counts()

anomaly_flag
False    40
Name: count, dtype: int64

In [110]:
df2[
    (df2['expense_ratio_pct'] < 0.1) |
    (df2['expense_ratio_pct'] > 2.5)
]

,amfi_code,scheme_name,fund_house,category,plan,return_1yr_pct,return_3yr_pct,return_5yr_pct,benchmark_3yr_pct,alpha,beta,sharpe_ratio,sortino_ratio,std_dev_ann_pct,max_drawdown_pct,aum_crore,expense_ratio_pct,morningstar_rating,risk_grade,anomaly_flag


In [112]:
df2.isnull().sum()

amfi_code             0
scheme_name           0
fund_house            0
category              0
plan                  0
return_1yr_pct        0
return_3yr_pct        0
return_5yr_pct        0
benchmark_3yr_pct     0
alpha                 0
beta                  0
sharpe_ratio          0
sortino_ratio         0
std_dev_ann_pct       0
max_drawdown_pct      0
aum_crore             0
expense_ratio_pct     0
morningstar_rating    0
risk_grade            0
anomaly_flag          0
dtype: int64

In [113]:
df2.to_csv(
    "../data/processed/scheme_performance_clean.csv",
    index=False
)

# Scheme Performance Data Cleaning

## Objective
Clean and validate mutual fund scheme performance metrics for analysis and database loading.

## Cleaning Steps Performed

- Reviewed dataset structure and column data types.
- Validated that return columns (`return_1yr_pct`, `return_3yr_pct`, `return_5yr_pct`) are numeric.
- Checked for missing values across all columns.
- Reviewed return distributions using descriptive statistics.
- Created an `anomaly_flag` column to identify potential abnormal return values.
- Validated `expense_ratio_pct` against the acceptable business range (0.1% – 2.5%).
- Verified data quality before exporting the cleaned dataset.

## Findings

- All return columns were already stored as numeric values.
- No missing values were found in key performance metrics.
- Return values were within expected ranges:
  - 1-Year Return: 4.26% to 24.93%
  - 3-Year Return: 5.14% to 23.39%
  - 5-Year Return: 5.43% to 23.80%
- No anomalous return values were identified.
- All expense ratios were within the expected range of 0.1%–2.5%.

## Output

Cleaned dataset saved as:

`data/processed/scheme_performance_clean.csv`

investor_transactions

In [115]:
df3=pd.read_csv(r"C:\Users\HP\OneDrive\Documents\mutual-fund-analytics\data\raw\08_investor_transactions - Copy.csv")
df3.head(5)

,investor_id,transaction_date,amfi_code,transaction_type,amount_inr,state,city,city_tier,age_group,gender,annual_income_lakh,payment_mode,kyc_status
0,INV003054,2024-01-01,119092,SIP,1834,Telangana,Hyderabad,T30,56+,Female,77.1,UPI,Verified
1,INV002952,2024-01-01,148567,Redemption,392882,Punjab,Amritsar,B30,18-25,Male,7.1,Cheque,Verified
2,INV003420,2024-01-01,118636,SIP,912,Haryana,Faridabad,B30,36-45,Male,47.2,Mandate,Verified
3,INV003436,2024-01-01,118634,SIP,1102,Maharashtra,Mumbai,T30,36-45,Female,54.4,Cheque,Pending
4,INV004691,2024-01-01,119094,Lumpsum,8682,Delhi,Noida,T30,26-35,Male,14.5,Net Banking,Pending


In [116]:
df3.shape

(32778, 13)

In [117]:
df3.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32778 entries, 0 to 32777
Data columns (total 13 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   investor_id         32778 non-null  object 
 1   transaction_date    32778 non-null  object 
 2   amfi_code           32778 non-null  int64  
 3   transaction_type    32778 non-null  object 
 4   amount_inr          32778 non-null  int64  
 5   state               32778 non-null  object 
 6   city                32778 non-null  object 
 7   city_tier           32778 non-null  object 
 8   age_group           32778 non-null  object 
 9   gender              32778 non-null  object 
 10  annual_income_lakh  32778 non-null  float64
 11  payment_mode        32778 non-null  object 
 12  kyc_status          32778 non-null  object 
dtypes: float64(1), int64(2), object(10)
memory usage: 3.3+ MB


In [118]:
df3.isnull().sum()

investor_id           0
transaction_date      0
amfi_code             0
transaction_type      0
amount_inr            0
state                 0
city                  0
city_tier             0
age_group             0
gender                0
annual_income_lakh    0
payment_mode          0
kyc_status            0
dtype: int64

In [119]:
df3.duplicated().sum()

np.int64(0)

In [120]:
df3.describe(include='all')

,investor_id,transaction_date,amfi_code,transaction_type,amount_inr,state,city,city_tier,age_group,gender,annual_income_lakh,payment_mode,kyc_status
count,32778,32778,32778.000000,32778,32778.000000,32778,32778,32778,32778,32778,32778.000000,32778,32778
unique,5000,516,NaN,3,NaN,12,24,2,5,2,NaN,4,2
top,INV001497,2025-04-28,NaN,SIP,NaN,Punjab,Kolkata,T30,26-35,Male,NaN,Net Banking,Verified
freq,12,88,NaN,19716,NaN,2965,2748,21719,13463,21809,NaN,8250,30146
mean,NaN,NaN,120264.617518,NaN,107437.318628,NaN,NaN,NaN,NaN,NaN,26.181219,NaN,NaN
std,NaN,NaN,14370.205345,NaN,150415.905084,NaN,NaN,NaN,NaN,NaN,20.805425,NaN,NaN
min,NaN,NaN,100016.000000,NaN,400.000000,NaN,NaN,NaN,NaN,NaN,3.000000,NaN,NaN
25%,NaN,NaN,118632.000000,NaN,3153.000000,NaN,NaN,NaN,NaN,NaN,10.600000,NaN,NaN
50%,NaN,NaN,119551.000000,NaN,17782.500000,NaN,NaN,NaN,NaN,NaN,19.700000,NaN,NaN
75%,NaN,NaN,120843.000000,NaN,189324.250000,NaN,NaN,NaN,NaN,NaN,37.400000,NaN,NaN


In [121]:
df3['transaction_type'].value_counts()

transaction_type
SIP           19716
Lumpsum        8095
Redemption     4967
Name: count, dtype: int64

### Transaction Type Validation

The transaction_type column was reviewed and contains only the expected values:

- SIP
- Lumpsum
- Redemption

No standardization was required.

In [122]:
df3['transaction_date'] = pd.to_datetime(
    df3['transaction_date']
)

In [123]:
df3.dtypes

investor_id                   object
transaction_date      datetime64[ns]
amfi_code                      int64
transaction_type              object
amount_inr                     int64
state                         object
city                          object
city_tier                     object
age_group                     object
gender                        object
annual_income_lakh           float64
payment_mode                  object
kyc_status                    object
dtype: object

In [126]:
print("Valid records:", (df3['amount_inr'] > 0).sum())
print("Invalid records:", (df3['amount_inr'] <= 0).sum())

Valid records: 32778
Invalid records: 0


In [127]:
df3['kyc_status'].value_counts()

kyc_status
Verified    30146
Pending      2632
Name: count, dtype: int64

In [128]:
df3['kyc_status'].unique()

array(['Verified', 'Pending'], dtype=object)

In [129]:
df3.to_csv(
    "../data/processed/investor_transactions_clean.csv",
    index=False
)

# Investor Transactions Data Cleaning

## Objective

Clean and validate investor transaction data before loading into the analytics database.

## Cleaning Steps Performed

- Reviewed dataset structure and column data types.
- Checked for missing values and duplicate records.
- Validated transaction_type values.
- Converted transaction_date to datetime format.
- Validated transaction amounts against the business rule: amount_inr > 0.
- Reviewed KYC status enum values.
- Prepared dataset for database loading.

## Findings

### Transaction Type Validation

The transaction_type column contained only the expected values:

- SIP
- Lumpsum
- Redemption

No standardization was required.

### Date Format Validation

The transaction_date column was successfully converted to datetime format.

### Amount Validation

Validated transaction amounts against the business rule:

amount_inr > 0

No invalid transaction amounts were found.

All 32,778 records passed validation.

### KYC Status Validation

The kyc_status column contained only valid enum values:

- Verified (30,146 records)
- Pending (2,632 records)

No invalid or inconsistent KYC status values were identified.

## Output

Cleaned dataset saved as:

`data/processed/investor_transactions_clean.csv`

In [4]:
import os

print(os.listdir("../data/processed"))

['fund_master_clean.csv', 'investor_transactions_clean.csv', 'nav_history_clean.csv', 'scheme_performance_clean.csv']
